NoClassDefFoundError in ElasticsearchRepository.findAllById

- version of `spring-data-elasticsearch`: `5.0.0-M5`
- version of `elasticsearch-java`: `8.3.1`

When calling [`findAllById`](https://github.com/spring-projects/spring-data-elasticsearch/blob/main/src/main/java/org/springframework/data/elasticsearch/repository/support/SimpleElasticsearchRepository.java#L150) of a minimal example application using release `5.0.0-M5` of `spring-data-elasticsearch`, the following error occurs:

```
jakarta.servlet.ServletException: Handler dispatch failed: java.lang.NoClassDefFoundError: org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder$Field
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1082)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:960)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1003)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:895)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:500)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:880)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1630)
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1600)
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1600)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1600)
        at org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:170)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1600)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:506)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:578)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1571)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1375)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:463)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1544)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1297)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.Server.handle(Server.java:562)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
        at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NoClassDefFoundError: org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder$Field
        at org.springframework.data.elasticsearch.client.erhlc.NativeSearchQueryBuilder.build(NativeSearchQueryBuilder.java:279)
        at org.springframework.data.elasticsearch.repository.support.SimpleElasticsearchRepository.getIdQuery(SimpleElasticsearchRepository.java:342)
        at org.springframework.data.elasticsearch.repository.support.SimpleElasticsearchRepository.findAllById(SimpleElasticsearchRepository.java:155)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at org.springframework.data.repository.core.support.RepositoryMethodInvoker$RepositoryFragmentMethodInvoker.lambda$new$0(RepositoryMethodInvoker.java:289)
        at org.springframework.data.repository.core.support.RepositoryMethodInvoker.doInvoke(RepositoryMethodInvoker.java:137)
        at org.springframework.data.repository.core.support.RepositoryMethodInvoker.invoke(RepositoryMethodInvoker.java:121)
        at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:516)
        at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:285)
        at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:628)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
        at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.doInvoke(QueryExecutorMethodInterceptor.java:160)
        at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.invoke(QueryExecutorMethodInterceptor.java:135)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
        at jdk.proxy2/jdk.proxy2.$Proxy74.findAllById(Unknown Source)
        at es8test.api.FooController.list(FooController.java:58)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067)
        ... 49 more
Caused by: java.lang.ClassNotFoundException: org.elasticsearch.search.fetch.subphase.highlight.HighlightBuilder$Field
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
```

---

The repository in question literally does nothing fancy at all and rather looks like this:

```java
package es8test.es.repo;

import org.springframework.data.elasticsearch.repository.ElasticsearchRepository;
import es8test.es.model.FooEntity;

public interface FooRepo extends ElasticsearchRepository<FooEntity, String> {}
```
---

The calling Spring `Controller` component method in the stacktrace above is quite simple as well, but any attempt at calling `findAllByIds` will fail:

```java
    @RequestMapping(value = "/list", method=RequestMethod.GET)
    public ResponseEntity<List<FooEntity>> list(@RequestParam List<String> ids) {
        List<FooEntity> results = StreamSupport.stream(
            repo.findAllById(ids).spliterator(), false
        ).toList();
        return new ResponseEntity<>(
            results, HttpStatus.CREATED
        );
    }
```

The `ApplicationRunner` app class is annotated with `@EnableAutoConfiguration` (doesn't matter if `ElasticsearchDataAutoConfiguration.class` is being excluded or not).

---

Fortunately, as a simple workaround one can just use a tactical `@Query` annotation for now:

```java
public interface FooRepo extends ElasticsearchRepository<FooEntity, String> {

    @Override
    @Query("{\"ids\": {\"values\": ?0 }}")
    Iterable<FooEntity> findAllById(Iterable<String> ids);

}
```

--- a/src/main/java/org/springframework/data/elasticsearch/repository/support/SimpleElasticsearchRepository.java
+++ b/src/main/java/org/springframework/data/elasticsearch/repository/support/SimpleElasticsearchRepository.java
@@ -15,8 +15,6 @@
  */
 package org.springframework.data.elasticsearch.repository.support;
 
-import static org.springframework.util.CollectionUtils.*;
-
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
@@ -28,11 +26,9 @@ import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.PageRequest;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.domain.Sort;
-import org.springframework.data.elasticsearch.client.erhlc.NativeSearchQueryBuilder;
 import org.springframework.data.elasticsearch.core.AbstractElasticsearchTemplate;
 import org.springframework.data.elasticsearch.core.ElasticsearchOperations;
 import org.springframework.data.elasticsearch.core.IndexOperations;
-import org.springframework.data.elasticsearch.core.MultiGetItem;
 import org.springframework.data.elasticsearch.core.RefreshPolicy;
 import org.springframework.data.elasticsearch.core.SearchHit;
 import org.springframework.data.elasticsearch.core.SearchHitSupport;
@@ -152,24 +148,12 @@ public class SimpleElasticsearchRepository<T, ID> implements ElasticsearchReposi
 		Assert.notNull(ids, "ids can't be null.");
 
 		List<T> result = new ArrayList<>();
-		Query idQuery = getIdQuery(ids);
-		if (isEmpty(idQuery.getIds())) {
-			return result;
-		}
+		Query query = getIdQuery(ids);
 
-		List<MultiGetItem<T>> multiGetItems = execute(
-				operations -> operations.multiGet(idQuery, entityClass, getIndexCoordinates()));
-
-		if (multiGetItems != null) {
-			multiGetItems.forEach(multiGetItem -> {
-
-				if (multiGetItem.hasItem()) {
-					result.add(multiGetItem.getItem());
-				}
-			});
-		}
-
-		return result;
+		List<SearchHit<T>> searchHitList = execute(
+				operations -> operations.search(query, entityClass, getIndexCoordinates()).getSearchHits());
+		// noinspection ConstantConditions
+		return (List<T>) SearchHitSupport.unwrapSearchHits(searchHitList);
 	}
 
 	@Override
@@ -339,7 +323,7 @@ public class SimpleElasticsearchRepository<T, ID> implements ElasticsearchReposi
 	private Query getIdQuery(Iterable<? extends ID> ids) {
 		List<String> stringIds = stringIdsRepresentation(ids);
 
-		return new NativeSearchQueryBuilder().withIds(stringIds).build();
+		return operations.idsQuery(stringIds);
 	}
 	// endregion
 

